src/Microsoft.Health.Fhir.Api/Features/HealthCheck/HealthCheckCachingMiddleware.cs

```diff
@@ -0,0 +1,108 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Fhir.Core;
+
+namespace Microsoft.Health.Fhir.Api.Features.HealthCheck
+{
+    /// <summary>
+    /// This is not the same as HealthCheckOptions.AllowCachingResponses.
+    /// This Middleware provides short term caching to protect against DoS attacks on the health check endpoint.
+    /// </summary>
+    internal sealed class HealthCheckCachingMiddleware : IDisposable
```

Comment:
Is there any reason we are not caching the result in `CosmosHealthCheck` and `SqlServerHealthCheck`?

Reply:
I also had an implementation for this approach. Middleware seemed like a better catch-all for all health checks as the goal is to add end point protection.

For consideration this was my implementation for caching the check itself:
```
using System;
using System.Threading;
using System.Threading.Tasks;
using EnsureThat;
using Microsoft.Extensions.Diagnostics.HealthChecks;

namespace Microsoft.Health.Fhir.Core
{
    public class CachedHealthCheck<T> : IHealthCheck
        where T : class, IHealthCheck
    {
        private readonly T _healthCheck;
        private DateTimeOffset _lastChecked;
        private HealthCheckResult _lastResult;
        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);

        public CachedHealthCheck(T healthCheck)
        {
            EnsureArg.IsNotNull(healthCheck, nameof(healthCheck));

            _healthCheck = healthCheck;
        }

        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken)
        {
            if (_lastChecked >= Clock.UtcNow.AddSeconds(-1))
            {
                return _lastResult;
            }

            await _semaphore.WaitAsync(cancellationToken);
            try
            {
                if (_lastChecked >= Clock.UtcNow.AddSeconds(-1))
                {
                    return _lastResult;
                }

                _lastResult = await _healthCheck.CheckHealthAsync(context, cancellationToken);
                _lastChecked = Clock.UtcNow;

                return _lastResult;
            }
            finally
            {
                _semaphore.Release();
            }
        }
    }
}
```

Reply:
If we are converned about CPU cycles serializing the response, the /metadata endpoint will be a lot heavier. Perhaps output caching should be extended to it (with a longer ttl)

Reply:
I was thinking of addressing metadata in as part of this https://github.com/microsoft/fhir-server/compare/personal/bkowitz/adds-aspnetcore-responsecompression?expand=1

Reply:
@jackliums @johnstairs in your opinion(s) do you feed its cleaner to go with the above `CachedHealthCheck<T>` approach?

Reply:
Updated